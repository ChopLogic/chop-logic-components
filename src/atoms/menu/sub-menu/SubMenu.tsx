import React, { PropsWithChildren, useRef } from 'react';
import { ChopLogicIcon } from '@atoms';
import { ChopLogicIconName, ChopLogicOrientationMode } from '@enums';
import { useClickOutside } from '@hooks';
import { ChopLogicMenuItem } from '@models';
import styles from './SubMenu.module.scss';

type Props = PropsWithChildren & {
  item: ChopLogicMenuItem;
  isSubMenuOpened: boolean;
  mode: ChopLogicOrientationMode;
  toggleSubMenu: () => void;
  closeSubMenu: () => void;
  openSubMenu: () => void;
  openedOn?: 'hover' | 'click';
};

export const SubMenu: React.FC<Props> = ({ item, isSubMenuOpened, toggleSubMenu, closeSubMenu, openSubMenu, openedOn, mode, children }) => {
  const { icon, link, label } = item;
  const ref = useRef(null);
  const dependentRef = useRef(null);

  const itemContent = link ? (
    <a href={link} target='_blank' rel='noreferrer'>
      <ChopLogicIcon name={icon} />
      {label}
    </a>
  ) : (
    <span>
      <ChopLogicIcon name={icon} />
      {label}
    </span>
  );

  const handleKeyDown = (e: React.KeyboardEvent<HTMLLIElement>) => {
    e.stopPropagation();
    if (e.key === ' ') {
      toggleSubMenu();
    }
  };

  const onClickOutsideHandler = () => {
    if (mode === ChopLogicOrientationMode.Vertical) return;
    closeSubMenu();
  };

  useClickOutside({ ref, dependentRef, onClickOutsideHandler });

  return (
    <li
      className={styles.subMenu}
      tabIndex={0}
      role='menuitem'
      aria-haspopup='true'
      aria-expanded={isSubMenuOpened}
      onKeyDown={handleKeyDown}
      onMouseEnter={openedOn === 'hover' ? openSubMenu : undefined}
      onMouseLeave={openedOn === 'hover' ? closeSubMenu : undefined}
      ref={ref}
    >
      <span className={styles.text} ref={dependentRef} onClick={toggleSubMenu}>
        {itemContent}
        <ChopLogicIcon name={isSubMenuOpened ? ChopLogicIconName.ArrowUp : ChopLogicIconName.ArrowDown} />
      </span>
      {isSubMenuOpened && children}
    </li>
  );
};
