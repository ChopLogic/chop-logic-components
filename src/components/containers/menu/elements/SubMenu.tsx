import React, { PropsWithChildren } from 'react';
import { ChopLogicIcon } from '@elements';
import { ChopLogicIconName } from '@enums';
import { ChopLogicMenuItem, ChopLogicTheme } from '@models';

import { StyledSubMenu, StyledSubMenuText } from '../Meny.styled.ts';

type SubMenuProps = PropsWithChildren & {
  item: ChopLogicMenuItem;
  theme: ChopLogicTheme;
  isSubMenuOpened: boolean;
  openedOn: 'click' | 'hover';
  toggleSubMenu: () => void;
  openSubMenu: () => void;
};

const SubMenu: React.FC<SubMenuProps> = ({ item, theme, isSubMenuOpened, toggleSubMenu, openSubMenu, openedOn, children }) => {
  const { icon, link, label } = item;

  const itemContent = link ? (
    <a href={link} target='_blank' rel='noreferrer'>
      <ChopLogicIcon name={icon} />
      {label}
    </a>
  ) : (
    <span>
      <ChopLogicIcon name={icon} />
      {label}
    </span>
  );

  const handleKeyDown = (e: React.KeyboardEvent<HTMLLIElement>) => {
    e.stopPropagation();
    if (e.key === ' ') {
      toggleSubMenu();
    }
  };

  return (
    <StyledSubMenu
      tabIndex={0}
      role='menuitem'
      aria-haspopup='true'
      aria-expanded={isSubMenuOpened}
      $theme={theme}
      $isOpened={isSubMenuOpened}
      onKeyDown={handleKeyDown}
    >
      <StyledSubMenuText $theme={theme} onClick={toggleSubMenu} onMouseEnter={openedOn === 'hover' ? openSubMenu : undefined}>
        {itemContent}
        <ChopLogicIcon name={isSubMenuOpened ? ChopLogicIconName.ArrowUp : ChopLogicIconName.ArrowDown} />
      </StyledSubMenuText>
      {isSubMenuOpened && children}
    </StyledSubMenu>
  );
};

export default SubMenu;
