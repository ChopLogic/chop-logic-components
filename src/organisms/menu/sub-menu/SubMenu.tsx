import { Icon } from '@atoms';
import { IconName, OrientationMode } from '@enums';
import { useClickOutside } from '@hooks';
import type { MenuItem } from '@models';
import { type FC, type KeyboardEvent, type PropsWithChildren, useRef } from 'react';

import styles from './SubMenu.module.scss';

type Props = PropsWithChildren & {
  item: MenuItem;
  isSubMenuOpened: boolean;
  mode: OrientationMode;
  toggleSubMenu: () => void;
  closeSubMenu: () => void;
  openSubMenu: () => void;
  openedOn?: 'hover' | 'click';
};

export const SubMenu: FC<Props> = ({
  item,
  isSubMenuOpened,
  toggleSubMenu,
  closeSubMenu,
  openSubMenu,
  openedOn,
  mode,
  children,
}) => {
  const { icon, link, label } = item;
  const ref = useRef(null);
  const dependentRef = useRef(null);

  const itemContent = link ? (
    <a href={link} target="_blank" rel="noreferrer">
      <Icon name={icon} />
      {label}
    </a>
  ) : (
    <span>
      <Icon name={icon} />
      {label}
    </span>
  );

  const handleKeyDown = (e: KeyboardEvent<HTMLElement>) => {
    e.stopPropagation();
    if (e.key === ' ') {
      toggleSubMenu();
    }
  };

  const onClickOutsideHandler = () => {
    if (mode === OrientationMode.Vertical) return;
    closeSubMenu();
  };

  useClickOutside({ ref, dependentRef, onClickOutsideHandler });

  return (
    <li
      className={styles.subMenu}
      tabIndex={0}
      role="menuitem"
      aria-haspopup="true"
      aria-expanded={isSubMenuOpened}
      onKeyDown={handleKeyDown}
      onMouseEnter={openedOn === 'hover' ? openSubMenu : undefined}
      onMouseLeave={openedOn === 'hover' ? closeSubMenu : undefined}
      ref={ref}
    >
      <span
        className={styles.text}
        ref={dependentRef}
        onClick={toggleSubMenu}
        onKeyDown={handleKeyDown}
      >
        {itemContent}
        <Icon name={isSubMenuOpened ? IconName.ArrowUp : IconName.ArrowDown} />
      </span>
      {isSubMenuOpened && children}
    </li>
  );
};
