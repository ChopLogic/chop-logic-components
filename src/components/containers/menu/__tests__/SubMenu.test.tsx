import { ChopLogicIconName, ChopLogicOrientationMode } from '@enums';
import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import SubMenu from '../elements/SubMenu';

import { LIGHT_THEME } from '@css';

vi.mock('@hooks', () => ({
  useClickOutside: vi.fn(),
}));

describe('SubMenu', () => {
  const mockToggleSubMenu = vi.fn();
  const mockCloseSubMenu = vi.fn();

  const itemWithLink = {
    id: '1',
    label: 'Item with Link',
    link: 'https://example.com',
    icon: ChopLogicIconName.ArrowUp,
  };

  const itemWithoutLink = {
    id: '2',
    label: 'Item without Link',
    icon: ChopLogicIconName.ArrowDown,
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <SubMenu
        item={itemWithoutLink}
        theme={LIGHT_THEME}
        isSubMenuOpened={true}
        toggleSubMenu={mockToggleSubMenu}
        closeSubMenu={mockCloseSubMenu}
        mode={ChopLogicOrientationMode.Horizontal}
      >
        <div>Submenu Content</div>
      </SubMenu>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with a link', () => {
    render(
      <SubMenu
        item={itemWithLink}
        theme={LIGHT_THEME}
        isSubMenuOpened={false}
        toggleSubMenu={mockToggleSubMenu}
        closeSubMenu={mockCloseSubMenu}
        mode={ChopLogicOrientationMode.Horizontal}
      />,
    );

    const link = screen.getByRole('link', { name: /item with link/i });
    expect(link).toHaveAttribute('href', 'https://example.com');
  });

  it('renders correctly without a link', () => {
    render(
      <SubMenu
        item={itemWithoutLink}
        theme={LIGHT_THEME}
        isSubMenuOpened={false}
        toggleSubMenu={mockToggleSubMenu}
        closeSubMenu={mockCloseSubMenu}
        mode={ChopLogicOrientationMode.Horizontal}
      />,
    );

    const span = screen.getByText(/item without link/i);
    expect(span).toBeInTheDocument();
  });

  it('calls toggleSubMenu on click', () => {
    render(
      <SubMenu
        item={itemWithoutLink}
        theme={LIGHT_THEME}
        isSubMenuOpened={false}
        toggleSubMenu={mockToggleSubMenu}
        closeSubMenu={mockCloseSubMenu}
        mode={ChopLogicOrientationMode.Horizontal}
      />,
    );

    const submenuText = screen.getByText(/item without link/i);
    fireEvent.click(submenuText);

    expect(mockToggleSubMenu).toHaveBeenCalledTimes(1);
  });

  it('toggles submenu on Space key press', () => {
    render(
      <SubMenu
        item={itemWithoutLink}
        theme={LIGHT_THEME}
        isSubMenuOpened={false}
        toggleSubMenu={mockToggleSubMenu}
        closeSubMenu={mockCloseSubMenu}
        mode={ChopLogicOrientationMode.Horizontal}
      />,
    );

    const submenu = screen.getByRole('menuitem');
    fireEvent.keyDown(submenu, { key: ' ' });

    expect(mockToggleSubMenu).toHaveBeenCalledTimes(2);
  });

  it('renders children when isSubMenuOpened is true', () => {
    render(
      <SubMenu
        item={itemWithoutLink}
        theme={LIGHT_THEME}
        isSubMenuOpened={true}
        toggleSubMenu={mockToggleSubMenu}
        closeSubMenu={mockCloseSubMenu}
        mode={ChopLogicOrientationMode.Horizontal}
      >
        <div>Submenu Content</div>
      </SubMenu>,
    );

    expect(screen.getByText(/submenu content/i)).toBeInTheDocument();
  });
});
