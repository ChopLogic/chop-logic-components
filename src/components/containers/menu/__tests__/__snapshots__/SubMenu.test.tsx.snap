// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubMenu > should match the snapshot 1`] = `
<DocumentFragment>
  <li
    aria-expanded="true"
    aria-haspopup="true"
    class="sc-ghWlax gRgehL"
    role="menuitem"
    tabindex="0"
  >
    <span
      class="sc-kLhKbu elIGzD"
    >
      <span>
        <span
          aria-label="SVG Icon Fallback"
          style="font-size: 1rem;"
        >
          ⬜
        </span>
        Item without Link
      </span>
      <span
        aria-label="SVG Icon Fallback"
        style="font-size: 1rem;"
      >
        ⬜
      </span>
    </span>
    <div>
      Submenu Content
    </div>
  </li>
</DocumentFragment>
`;
